### PR TITLE
Real-World Angular Series - Part 4

### PR DESCRIPTION
**Pull request by kmaida**
Publish date: July 6

### CODE REVIEW

File: _posts/2017-07-06-real-world-angular-series-part-4.markdown [hunk 1]
```diff
@@ -0,0 +1,880 @@
+---
+layout: post
+title: "Real-World Angular Series - Part 4: Access Management, Admin, and Detail Pages"
+description: "Build and deploy a real-world app with MongoDB, Express, Angular, and Node (MEAN): access management, admin listing, and detail pages."
```
Comment (kukicado) [inline _posts/2017-07-06-real-world-angular-series-part-4.markdown]: admin listing reads a little weird. Maybe "admin dashboard" or something similar. WDYT?

File: _posts/2017-07-06-real-world-angular-series-part-4.markdown [hunk 2]
```diff
@@ -0,0 +1,880 @@
...
+- express
+- mean
+related:
+- 2017-06-27-real-world-angular-series-part-1
+- 2017-06-29-real-world-angular-series-part-2
+- 2017-07-04-real-world-angular-series-part-3
+
+---
+
+**TL;DR:** This tutorial series covers building and deploying a full-stack JavaScript application from the ground up with hosted [MongoDB](https://www.mongodb.com/), [Express](https://expressjs.com/), [Angular (v2+)](https://angular.io), and [Node.js](https://nodejs.org) (also known as MEAN stack). The completed code is available in the [mean-rsvp-auth0 GitHub repo](https://github.com/auth0-blog/mean-rsvp-auth0/). **Part 4 of the tutorial series covers access management with Angular, displaying admin data, and setting up detail pages with tabs.**
```
Comment (kukicado) [inline _posts/2017-07-06-real-world-angular-series-part-4.markdown]: also known as the MEAN stack

File: _posts/2017-07-06-real-world-angular-series-part-4.markdown [hunk 3]
```diff
@@ -0,0 +1,880 @@
...
+1. Is the user authenticated?
+2. Does the authenticated user have admin privileges?
+
+{% include tweet_quote.html quote_text="Angular route guards determine whether a user should be allowed to access a specific route." %}
+
+### Redirecting To and From Login
+
+There will be an important extra feature in our authentication route guard that will require some updates to our authentication service: _redirection to and from login_. When an unauthenticated user arrives at our app, we want to _limit_ disruptions to their experience as much as possible. This means that we'll surface links to authenticated routes in our _public_ event listing, and then prompt the user to log in when such links are clicked. After they authenticate, they'll be redirected to the protected route.
+
+This reduces friction in the application because we're taking care not to redirect users back to a homepage. This would force them to try to find their own way back to the route they were originally trying to access. It also helps keep the user on track if they manually typed in a URL or clicked a link someone else sent them.
```
Comment (kukicado) [inline _posts/2017-07-06-real-world-angular-series-part-4.markdown]: Love this!

File: _posts/2017-07-06-real-world-angular-series-part-4.markdown [hunk 4]
```diff
@@ -0,0 +1,880 @@
...
+  providers: [
+    AuthGuard,
+    AdminGuard
+  ],
+  ...
+})
+...
+```
+
+We'll import our two new route guards and then add them to the `providers` array.
```
Comment (kukicado) [inline _posts/2017-07-06-real-world-angular-series-part-4.markdown]: two new => I don't think you need "new" here since these are the first and only route guards you're building.

File: _posts/2017-07-06-real-world-angular-series-part-4.markdown [hunk 5]
```diff
@@ -0,0 +1,880 @@
...
+  }
+
+  ngOnDestroy() {
+    this.eventsSub.unsubscribe();
+  }
+
+}
+```
+
+This is very similar to the Home component we set up in <a href="https://auth0.com/blog/real-world-angular-series-part-3#angular-home-events">Angular: Home Component Event List</a>. The only difference is that we'll import the Auth service. Other than that, we'll set the title, get the full admin events list, implement search functionality, and unsubscribe on destruction of the component.
```
Comment (kukicado) [inline _posts/2017-07-06-real-world-angular-series-part-4.markdown]: Other than that, we'll set the title, get the full admin events list, implement search functionality, and unsubscribe on destruction of the component => Other than that, we'll set the title, get the full admin events list, implement search functionality, and unsubscribe on destruction of the component just like we did in the Home component.  

WDYT - it feels like the sentence ended abruptly?

File: _posts/2017-07-06-real-world-angular-series-part-4.markdown [hunk 6]
```diff
@@ -0,0 +1,880 @@
...
+  <!-- Error loading events -->
+  <p *ngIf="error" class="alert alert-danger">
+    <strong>Oops!</strong> There was an error retrieving event data.
+  </p>
+
+</ng-template>
+{% endraw %}
+{% endhighlight %}
+
+Again, this is very similar to our Home component's implementation. However, we'll start with a paragraph greeting our admin user.We're showing icons in our event list indicating if an event is in the past (with the calendar icon) or if it has `viewPublic: false`. Also, we're only linking the title of the event to its detail page instead of the entire list item because we'll be adding "Edit" and "Delete" buttons to each event later.
```
Comment (kukicado) [inline _posts/2017-07-06-real-world-angular-series-part-4.markdown]: users.We're => users. We're (needs space)

File: _posts/2017-07-06-real-world-angular-series-part-4.markdown [hunk 7]
```diff
@@ -0,0 +1,880 @@
...
+}
+```
+
+Because our Admin component and API route are protected, you'll have to log into the app with the admin user you specified in the <a href="https://auth0.com/blog/real-world-angular-series-part-2#admin-authorization">Admin Authorization</a> section of Part 2 in order to view the page. Once you're logged in, the Admin component should look something like this:
+
+![Angular admin page](https://cdn.auth0.com/blog/mean-series/admin-list.jpg)
+
+If unauthenticated users attempt to access this page, they'll be prompted to log in. If they are admin upon logging in, they'll be granted access. If they don't have admin rights, they'll be redirected to the homepage by our admin route guard. If a user logs out from this page, they'll also be redirected to the homepage. Try it out!
+
+> **Security Note:** Even if a user was somehow able to circumvent the front end protection, the Node API would not return the events data without the correct admin role concealed in the opaque access token.
```
Comment (kukicado) [inline _posts/2017-07-06-real-world-angular-series-part-4.markdown]: I thought only the id_token was opaque or at least we referred to it as opaque while the access_token is not? Can you confirm?

File: _posts/2017-07-06-real-world-angular-series-part-4.markdown [hunk 8]
```diff
@@ -0,0 +1,880 @@
...
+    this.tabSub.unsubscribe();
+    this.eventSub.unsubscribe();
+  }
+
+}
+```
+
+As always, first we'll add our imports. Let's dive right into the class, covering the imports as we go through the code.
+
+This time, we won't set a `pageTitle` immediately. We first need to retrieve the event data from the API. We'll also grab the event's `id` by subscribing to the [`ActivatedRoute`](https://angular.io/api/router/ActivatedRoute) route parameters observable. We'll subscribe to the route's query parameters observable to set the `tab`. As usual, we'll get our event data from the API service, annotating results with the `EventModel` type. For this component and its children, we also want to know if the event is has already ended so we'll use an `eventPast` property to track this with the `eventPast()` method we added to our `UtilsService` in the <a href="https://auth0.com/blog/real-world-angular-series-part-3#angular-utility-service" target="_self">Angular: Create a Utility Service</a> section of Part 3.
```
Comment (kukicado) [inline _posts/2017-07-06-real-world-angular-series-part-4.markdown]: grab the event's => grab the events

File: _posts/2017-07-06-real-world-angular-series-part-4.markdown [hunk 9]
```diff
@@ -0,0 +1,880 @@
...
+
+![Angular MEAN app - event detail component](https://cdn.auth0.com/blog/mean-series/event-details.jpg)
+
+Now that we have our event details, we're ready to implement the logic to manage RSVPs next time.
+
+---
+
+## Summary
+
+In Part 4 of our Real-World Angular Series, we've covered access management with Angular, displaying admin data, and setting up detail pages with tabs. In the next part of the tutorial series, we'll tackle simple animation as well as creating and updating data with a template-driven form.
```
Comment (kukicado) [inline _posts/2017-07-06-real-world-angular-series-part-4.markdown]: Will you also cover reactive / data-driven froms - I think they are more applicable to Angular 2 development.

File: _posts/2017-07-06-real-world-angular-series-part-4.markdown (hunk 1 above)
Comment (kmaida) [inline _posts/2017-07-06-real-world-angular-series-part-4.markdown]: It isn't the admin dashboard yet though, it's an admin-authorized listing of events. Because the dashboard comes later, I want to be wary of being misleading. I'll change it to something else. :)

File: _posts/2017-07-06-real-world-angular-series-part-4.markdown (hunk 2 above)
Comment (kmaida) [inline _posts/2017-07-06-real-world-angular-series-part-4.markdown]: Leaving this as-is, for reasons cited in previous PR. :)

File: _posts/2017-07-06-real-world-angular-series-part-4.markdown (hunk 4 above)
Comment (kmaida) [inline _posts/2017-07-06-real-world-angular-series-part-4.markdown]: 👍

File: _posts/2017-07-06-real-world-angular-series-part-4.markdown (hunk 5 above)
Comment (kmaida) [inline _posts/2017-07-06-real-world-angular-series-part-4.markdown]: I worry that adding more to this sentence makes it very long / too wordy, especially when it was already stated that we're doing the same thing as we did in the Home component.

File: _posts/2017-07-06-real-world-angular-series-part-4.markdown (hunk 6 above)
Comment (kmaida) [inline _posts/2017-07-06-real-world-angular-series-part-4.markdown]: 👍

File: _posts/2017-07-06-real-world-angular-series-part-4.markdown (hunk 7 above)
Comment (kmaida) [inline _posts/2017-07-06-real-world-angular-series-part-4.markdown]: The access token is the opaque one. The ID token is the one that the client can read and decode. :)

File: _posts/2017-07-06-real-world-angular-series-part-4.markdown (hunk 8 above)
Comment (kmaida) [inline _posts/2017-07-06-real-world-angular-series-part-4.markdown]: `event's` is correct here because it's referring to a possessive: the `event's ID`, not the "events" (plural).

File: _posts/2017-07-06-real-world-angular-series-part-4.markdown (hunk 9 above)
Comment (kmaida) [inline _posts/2017-07-06-real-world-angular-series-part-4.markdown]: Reactive/data-driven forms have almost an entire section of tutorial to themselves, since we do some complex reactive form group custom validation. :) Don't worry, they're covered in glorious detail later!

Comment (kukicado) [inline _posts/2017-07-06-real-world-angular-series-part-4.markdown]: Great to hear :D
